server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -876,61 +853,53 @@ private void cancelOfflineTableMigrations(KeyExtent extent) {
     }
   }
 
-  private void repairMetadata(Text row) {
-    Manager.log.debug("Attempting repair on {}", row);
-    // ACCUMULO-2261 if a dying tserver writes a location before its lock information propagates, it
-    // may cause duplicate assignment.
-    // Attempt to find the dead server entry and remove it.
+  private void logIncorrectTabletLocations(TabletMetadata tabletMetadata) {
     try {
       Map<Key,Value> future = new HashMap<>();
       Map<Key,Value> assigned = new HashMap<>();
-      KeyExtent extent = KeyExtent.fromMetaRow(row);
-      String table = AccumuloTable.METADATA.tableName();
-      if (extent.isMeta()) {
-        table = AccumuloTable.ROOT.tableName();
+      KeyExtent extent = tabletMetadata.getExtent();
+      var level = Ample.DataLevel.of(extent.tableId());
+
+      Stream<? extends Entry<Key,Value>> entries;
+      if (level == Ample.DataLevel.ROOT) {
+        RootTabletMetadata rtm = RootTabletMetadata.read(manager.getContext());
+        entries = rtm.getKeyValues();
+      } else {
+        Scanner scanner =
+            manager.getContext().createScanner(level.metaTable(), Authorizations.EMPTY);
+        scanner.fetchColumnFamily(CurrentLocationColumnFamily.NAME);
+        scanner.fetchColumnFamily(FutureLocationColumnFamily.NAME);
+        scanner.setRange(new Range(extent.toMetaRow()));
+        entries = scanner.stream();
```

Comment:
Could add an `onClose()` here to close the scanner. And then maybe close the stream after we create the maps.

or

Maybe a new method that returns the stream:

```java
  private Stream<? extends Entry<Key,Value>> getMetaEntries(KeyExtent extent)
      throws TableNotFoundException, InterruptedException, KeeperException {
    Ample.DataLevel level = Ample.DataLevel.of(extent.tableId());
    if (level == Ample.DataLevel.ROOT) {
      return RootTabletMetadata.read(manager.getContext()).getKeyValues();
    } else {
      Scanner scanner = manager.getContext().createScanner(level.metaTable(), Authorizations.EMPTY);
      scanner.fetchColumnFamily(CurrentLocationColumnFamily.NAME);
      scanner.fetchColumnFamily(FutureLocationColumnFamily.NAME);
      scanner.setRange(new Range(extent.toMetaRow()));
      return scanner.stream().onClose(scanner::close);
    }
  }
```

Then we could use the stream in a try-with-resources:

```java
      try (Stream<? extends Entry<Key,Value>> entries = getMetaEntries(extent)) {
        entries.forEach(entry -> {
          if (entry.getKey().getColumnFamily().equals(CurrentLocationColumnFamily.NAME)) {
            assigned.put(entry.getKey(), entry.getValue());
          } else if (entry.getKey().getColumnFamily().equals(FutureLocationColumnFamily.NAME)) {
            future.put(entry.getKey(), entry.getValue());
          }
        });
      }
```

I am not sure if either of these options are an improvement in terms of simplicity or readability though so might not be worth it.

Reply:
I like introducing the method, I wanted to close the stream but could not think of a clean way to do it.

Reply:
changed in [e0e2e92](https://github.com/apache/accumulo/pull/4668/commits/e0e2e929f95ab95d1f4dd2659dc2822edb9bcd7d)